modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/model/impl/ExtensionPointImpl.java

```diff
@@ -76,39 +95,114 @@ public String getSuperComponent() {
-            return contribs;
-        }
+    @Override
+    @SuppressWarnings("deprecation")
+    public void register(Extension extension) {
         // should compute now the contributions
         if (contributions != null) {
-            if (xmap == null) {
-                xmap = new XMap();
-                for (Class<?> contrib : contributions) {
-                    if (contrib != null) {
-                        xmap.register(contrib);
-                    } else {
-                        throw new RuntimeException("Unknown implementation class when contributing to "
-                                + owner.getComponent().getName());
-                    }
-                }
-            }
             try {
-                contribs = xmap.loadAll(new XMapContext(extension.getContext()), extension.getElement());
+                Context xctx = new XMapContext(extension.getContext());
+                // backward compatibility
+                if (extension.getContributions() == null) {
+                    // overload use case: loaded contributions should use the old descriptor so should not be
+                    // recomputed
+                    extension.setContributions(getXmap().loadAll(xctx, extension.getElement()));
+                }
+                // fill up registry
+                Registry registry = getRegistry();
+                String flag = extension.getId();
+                if (registry != null && !registry.isFlagged(flag)) {
+                    registry.flag(flag);
+                    getXmap().register(registry, xctx, extension.getElement(), flag);
+                }
             } catch (XMapException e) {
                 throw new RuntimeException(
                         e.getMessage() + " while processing component: " + extension.getComponent().getName().getName(),
                         e);
             }
-            extension.setContributions(contribs);
+        } else {
+            throw new RuntimeException(String.format(
+                    "Cannot contribute contributions from component '%s': extension point '%s:%s' is missing contribution classes",
+                    extension.getComponent().getName(), superComponent, name));
+        }
+    }
+
+    @Override
+    public void unregister(Extension extension) {
+        try {
+            getXmap().unregister(getRegistry(), extension.getId());
+        } catch (XMapException e) {
+            log.error(e.getMessage() + " while unprocessing component: " + extension.getComponent().getName().getName(),
+                    e);
         }
-        return contribs;
+    }
+
+    @Override
+    public Registry getRegistry() {
+        if (registry == null) {
+            registry = computeFinalRegistry();
+            if (registry == null) {
+                registry = NULL_REGISTRY;
+            }
+        }
+        if (registry.isNull()) {
+            return null;
```

Comment:
Why returning null instead of the `NULL_REGISTRY`?

Reply:
NULL_REGISTRY is only used for internal caching logics (to avoid detecting over and over that there is no registry defined), i can expose it and return it if better but i'm not convinced it's clearer (?)

Reply:
I think it is, the other side is getting NPE in border cases, it might help no?

Reply:
Let's see when moving this API to the ComponentManager then

Reply:
I re-arranged the logics while moving to ComponentManager, see #4537